external/fv3fit/fv3fit/pytorch/cyclegan/reloadable.py

```diff
@@ -170,7 +174,80 @@ def predict(self, X: xr.Dataset, reverse: bool = False) -> xr.Dataset:
             for i in range(0, tensor.shape[0], n_batch):
-                new: torch.Tensor = generator(tensor[i : i + n_batch])
+                new: torch.Tensor = generator(
+                    (time[i : i + n_batch], tensor[i : i + n_batch])
+                )
                 outputs[i : i + new.shape[0]] = new
         predicted = self.unpack_tensor(outputs, domain=output_domain)
+        predicted["time"] = X["time"]
         return predicted
+
+
+def _pack_to_tensor(
+    ds: xr.Dataset,
+    timesteps: int,
+    state_variables: Iterable[str],
+    scalers: Mapping[str, StandardScaler],
+) -> torch.Tensor:
+    """
+    Packs the dataset into a tensor to be used by the pytorch model.
+
+    Subdivides the dataset evenly into windows
+    of size (timesteps + 1) with overlapping start and end points.
+    Overlapping the window start and ends is necessary so that every
+    timestep (evolution from one time to the next) is included within
+    one of the windows.
+
+    Args:
+        ds: dataset containing values to pack
+        timesteps: number timesteps to include in each window after initial time
+
+    Returns:
+        tensor of shape [window, time, tile, x, y, feature]
+    """
+
+    expected_dims: Tuple[str, ...] = ("time", "tile", "x", "y")
+    if "z" in ds.dims:
+        expected_dims += ("z",)
+    ds = ds.transpose(..., *expected_dims)
+    if timesteps > 0:
+        n_times = ds.time.size
+        n_windows = int((n_times - 1) // timesteps)
+        # times need to be evenly divisible into windows
+        ds = ds.isel(time=slice(None, n_windows * timesteps + 1))
+    all_data = []
+    for varname in state_variables:
+        var_dims = ds[varname].dims
+        if tuple(var_dims[:4]) != expected_dims[:4]:
+            raise ValueError(
+                f"received variable {varname} with " f"unexpected dimensions {var_dims}"
+            )
+        data = ds[varname].values
+        normalized_data = scalers[varname].normalize(data)
+        if timesteps > 0:
+            # segment time axis into windows, excluding last time of each window
+            data = normalized_data[:-1, :].reshape(
+                n_windows, timesteps, *data.shape[1:]
+            )
+            # append first time of next window to end of each window
+            if n_windows > 1:
+                end_data = np.concatenate(
+                    [data[1:, :1, :], normalized_data[None, -1:, :]], axis=0
+                )
+            else:
+                end_data = normalized_data[None, -1:, :]
+            data = np.concatenate([data, end_data], axis=1)
+        else:
+            data = normalized_data
+        if "z" not in var_dims:
+            # need a z-axis for concatenation into feature axis
+            data = data[..., np.newaxis]
+        all_data.append(data)
+    concatenated_data = np.concatenate(all_data, axis=-1)
+    time = cftime.date2num(ds["time"].values, "seconds since 1970-01-01").astype(
```

Comment:
Would be nice if a lot of the time conversion and windowing logic could be written once instead of here and the `WindowedZarrLoader`. Is that possible?

Reply:
OK that's fine, just worth keeping an eye on repeated code as a possible source of bugs.